physical/raft/fsm.go

```diff
@@ -606,86 +609,36 @@ func (f *FSM) Restore(r io.ReadCloser) error {
 		return nil
 	}
 
-	snapMeta := r.(*boltSnapshotMetadataReader).Metadata()
-
-	protoReader := NewDelimitedReader(r, math.MaxInt32)
-	defer protoReader.Close()
+	snapshotInstaller, ok := r.(*boltSnapshotInstaller)
+	if !ok {
+		return errors.New("expected snapshot installer object")
+	}
 
 	f.l.Lock()
 	defer f.l.Unlock()
 
-	// Delete the existing data bucket and create a new one.
-	f.logger.Debug("snapshot restore: deleting bucket")
-	err := f.db.Update(func(tx *bolt.Tx) error {
-		err := tx.DeleteBucket(dataBucketName)
-		if err != nil {
-			return err
-		}
-
-		_, err = tx.CreateBucket(dataBucketName)
-		if err != nil {
-			return err
-		}
-
-		return nil
-	})
-	if err != nil {
-		f.logger.Error("could not restore snapshot: could not clear existing bucket", "error", err)
+	// Close the db file
+	if err := f.db.Close(); err != nil {
+		f.logger.Error("failed to close database file", "error", err)
 		return err
 	}
 
-	// If we are testing a failed snapshot error here.
-	if f.testSnapshotRestoreError {
-		return errors.New("Test error")
-	}
-
-	f.logger.Debug("snapshot restore: deleting bucket done")
-	f.logger.Debug("snapshot restore: writing keys")
+	dbPath := filepath.Join(f.path, databaseFilename)
 
-	var done bool
-	var keys int
-	for !done {
-		err := f.db.Update(func(tx *bolt.Tx) error {
-			b := tx.Bucket(dataBucketName)
-			s := new(pb.StorageEntry)
+	f.logger.Info("installing snapshot to FSM")
 
-			// Commit in batches of 50k. Bolt holds all the data in memory and
-			// doesn't split the pages until commit so we do incremental writes.
-			// This is safe since we have a write lock on the fsm's lock.
-			for i := 0; i < 50000; i++ {
-				err := protoReader.ReadMsg(s)
-				if err != nil {
-					if err == io.EOF {
-						done = true
-						return nil
-					}
-					return err
-				}
-
-				err = b.Put([]byte(s.Key), s.Value)
-				if err != nil {
-					return err
-				}
-				keys += 1
-			}
-
-			return nil
-		})
-		if err != nil {
-			f.logger.Error("could not restore snapshot", "error", err)
-			return err
-		}
-
-		f.logger.Trace("snapshot restore: writing keys", "num_written", keys)
+	// Install the new boltdb file
+	if err := snapshotInstaller.Install(dbPath); err != nil {
+		f.logger.Error("failed to install snapshot", "error", err)
+		return errwrap.Wrapf("failed to install snapshot database: {{err}}", err)
 	}
 
-	f.logger.Debug("snapshot restore: writing keys done")
+	f.logger.Info("snapshot installed")
 
-	// Write the metadata after we have applied all the snapshot data
-	f.logger.Debug("snapshot restore: writing metadata")
-	if err := f.witnessSnapshot(snapMeta); err != nil {
-		f.logger.Error("could not write metadata", "error", err)
-		return err
+	// Open the new file
```

Comment:
If at any point this method fails, the original db file is re-opening, correct?

Reply:
Good catch, take a look at aabe201 for the fix

Reply:
Looks good!